ntoskrnl/io/iomgr/deviface.c

```diff
@@ -304,7 +305,190 @@ IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlStringFromGUID() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    BaseKeyName.Length = 0;
+    BaseKeyName.MaximumLength += GuidString.Length;
+    BaseKeyName.Buffer = ExAllocatePoolWithTag(NonPagedPool,
+                                               BaseKeyName.MaximumLength,
+                                               TAG_IO);
+    if (!BaseKeyName.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto cleanup;
+    }
+
+    /* Concatenate alias GUID string with base key */
+    Status = RtlAppendUnicodeStringToString(&BaseKeyName, &GuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    /* Open base registry key */
+    Status = IoOpenDeviceInterfaceRegistryKey(&BaseKeyName,
+                                              GENERIC_READ,
+                                              &InterfaceKey);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("IoOpenDeviceInterfaceRegistryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    Status = ZwQueryKey(InterfaceKey,
+                        KeyFullInformation,
+                        NULL,
+                        0,
+                        &Size);
+    if (!NT_SUCCESS(Status) && Status != STATUS_BUFFER_TOO_SMALL)
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    FullInfo = ExAllocatePoolWithTag(PagedPool,
+                                     Size,
+                                     TAG_IO);
+    if (!FullInfo)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto cleanup;
+    }
+
+    Status = ZwQueryKey(InterfaceKey,
+                        KeyFullInformation,
+                        FullInfo,
+                        Size,
+                        &Size);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    /* Enumerate subkeys (i.e. the different device objects) */
+    for (i = 0; i < FullInfo->SubKeys; i++)
+    {
+        Status = ZwEnumerateKey(InterfaceKey,
+                                i,
+                                KeyBasicInformation,
+                                NULL,
+                                0,
+                                &Size);
+        if (Status == STATUS_NO_MORE_ENTRIES)
+        {
+            break;
+        }
+        else if (!NT_SUCCESS(Status) && Status != STATUS_BUFFER_TOO_SMALL)
+        {
+            DPRINT1("ZwEnumerateKey() failed with status 0x%08lx\n", Status);
+            goto cleanup;
+        }
+
+        BasicInfo = ExAllocatePoolWithTag(PagedPool,
```

Comment:
You seem to be leaking this. You allocate a new buffer on every iteration of the loop, but you only free the very last one.
You can also reuse the buffer (whenever the previous one is large enough) to avoid repeated allocation:
```
ULONG BasicInfoLength = 0;
PKEY_BASIC_INFORMATION BasicInfo = NULL;
ULONG NeededLength;

for (...)
{
    Status = ZwEnumerateKey(... BasicInfo, BasicInfoLength, &NeededLength);
    if (Status == STATUS_BUFFER_TOO_SMALL)
    {
        if (BasicInfo) ExFreePool(BasicInfo);
        BasicInfo = ExAllocatePool(NeededLength);
        BasicInfoLength = NeededLength;
        i--; continue; /* could use a while loop to avoid the i-- */
    }
}
```

Reply:
OK, I applied this suggestion (similarly to other code examples in this file), and now it seems no longer fail to free `AliasSymbolicLinkName` in sysaudio, so it's no longer leaking. But now the system fails with exactly the same BSOD 0x7F, same as before, without the implementation at all. :cry: And no any other suspicious debug prints from my code. Either it needs something another, or my one is still wrong somewhere.